src/rustdoc-json-types/lib.rs

```diff
@@ -96,28 +107,37 @@ pub struct Span {
     pub end: (usize, usize),
 }
 
+/// Information about the deprecation of an API entry.
```

Comment:
In the case of "calling convention" & "ABI", my idea was to use different synonyms for a concept to make them easier to understand if a person happens to better understand what's a calling convention than what's an ABI. I did the same thing with "bounds" & "constraints", I used to struggle with understanding Rust until I realised that a "bound" is simply a "constraint"

Reply:
And with generic "arguments" vs "parameters", the confusion is already in the names of the structs in the library, there is `GenericArg`, but also `GenericParamDef`, if we stick to 1 word, which should it be?

Reply:
> my idea was to use different synonyms

but ABI and calling conventions don't been quite the same thing. Both [the reference](https://doc.rust-lang.org/reference/items/functions.html#extern-function-qualifier) and [the book](https://doc.rust-lang.org/book/ch19-01-unsafe-rust.html?highlight=ABI#calling-rust-functions-from-other-languages) talk about it.

>  to make them easier to understand if a person happens to better understand what's a calling convention than what's an ABI

fine, but discuss that later. The comment on `abi: Abi`, and the first paragraph of `enum Abi` should both be in tearm of ABI/Application Binary Interace. If you want to discuss calling conventions later, only do that after talking about what it actually is.

Reply:
> And with generic "arguments" vs "parameters", the confusion is already in the names of the structs in the library, there is GenericArg, but also GenericParamDef, if we stick to 1 word, which should it be?

No, these are entirely different things. *Parameter* and *argument* (or equivalently *formal* and *actual* parameters) are well-established technical terms that have *precise* meanings.

> In the case of "calling convention" & "ABI", my idea was to use different synonyms for a concept to make them easier to understand if a person happens to better understand what's a calling convention than what's an ABI. I did the same thing with "bounds" & "constraints", I used to struggle with understanding Rust until I realised that a "bound" is simply a "constraint"

I appreciate the effort but I think we need to be a lot more careful here. Picking words and "synonyms" willy-nilly in the context of something as technical as a programming language hurts the quality of the docs in both the short and the long run. Indeed, "bound", "constraint", "restriction", "predicate", "clause" (to continue your example) may all refer to the same thing (natural languages are highly context-dependent, flexible and malleable after all) but as soon as we enter the world of jargons and technical language the rules change a bit and words that may have a relatively lax and general meaning may have a super specific meaning assigned to it inside a specific jargon. For example, take a look at "expression", "statement", "judgement", "type", "kind", "term", "value". Some of these words may be synonymous in everyday language but in the world of Rust, Haskell, programming language design or type theory these may refer to very concrete very precise things.

**I think the proper way to go about this would be to use the *precise* term as used by rustc and/or the Rust community and to optionally add a short explainer paragraph** (containing hyperlinks to more elaborate resources or that draws comparison to similar concepts etc.).

To go back to "bounds" and "constraints". In rustc, these terms are distinct. *Bounds* include *trait bounds* (`T: Trait`, `u32: for<'a> Trait<'a, Output = &'a ()>`) and outlives-bounds aka lifetime bounds (`T: 'a`, `'a: 'b`). Yes, in Haskell you would call that a "constraint" but in the Rust world you generally wouldn't (unless you're chatting with your peers where you likely won't speak in precise technical language) and `rustdoc_json_types` ought to be precise. `Output = &'a ()` would be a constraint OTOH, more specifically an associated item (equality) constraint where a constraint is either an equality or a *bound*. Furthermore, `ty::Clause` and `ty::Predicate` are distinct.

Reply:
And to clarify, my comment shouldn't be construed as being *prescriptive*. I'm very glad you've picked up on the daunting task of documenting the entirety this crate, that's very commendable!

Making use of well-established terminology is not a form of *gate-keeping* as long as we immediately define what we mean. It helps us establish a common language to talk about things in our environment. For one, it enables users of rustdoc-json-types to join the conversation of rustdoc and rustc devs, e.g., in issues, PRs, on Zulip etc.